model/qti/scale/ScaleStorageService.php

```diff
@@ -0,0 +1,216 @@
+
+        foreach ($scalesDir->getIterator() as $file) {
+            if (!$file instanceof File) {
+                continue;
+            }
+
+            if ($file->getMimeType() === 'application/json') {
+                try {
+                    $content = $file->read();
+                    $scaleData = json_decode($content, true, 512, JSON_THROW_ON_ERROR);
+                    $scales[sprintf('scales/%s', $file->getBasename())] = $scaleData;
+                } catch (JsonException $e) {
+                    // Skip files that cannot be decoded
+                    continue;
+                }
+            }
+        }
+
+        return $scales;
+    }
+
+    public function generateRelativePath(string $identifierOrPath): string
+    {
+        $baseName = preg_replace('/[^a-zA-Z0-9_-]+/', '_', $identifierOrPath);
+        $baseName = trim($baseName, '_');
+
+        if ($baseName === '') {
+            $baseName = 'scale';
+        }
+
+        return sprintf(
+            '%s/%s_%s.json',
+            self::SCALES_DIRECTORY,
+            strtolower($baseName),
+            substr(sha1(uniqid((string)mt_rand(), true)), 0, 8)
+        );
+    }
+
+    public function isScalePath(?string $value): bool
+    {
+        if ($value === null) {
+            return false;
+        }
+
+        $path = $this->sanitizeRelativePath($value);
+
+        return strpos($path, self::SCALES_DIRECTORY . '/') === 0 && substr($path, -5) === '.json';
+    }
+
+    private function encodePayload(array $payload): string
+    {
+        return json_encode(
+            $payload,
+            JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE | JSON_PRETTY_PRINT
+        );
+    }
+
+    private function ensureScalesDirectory(core_kernel_classes_Resource $item, string $relativePath): Directory
+    {
+        $directory = $this->getScalesDirectory($item);
+
+        if (!$directory->exists()) {
+            $directory->getFileSystem()->createDirectory($directory->getPrefix());
+        }
+
+        $subPath = trim(substr(dirname($relativePath), strlen(self::SCALES_DIRECTORY)), '/');
+        if ($subPath !== '' && $subPath !== '.') {
+            $subDirectory = $directory->getDirectory($subPath);
+            if (!$subDirectory->exists()) {
+                $subDirectory->getFileSystem()->createDirectory($subDirectory->getPrefix());
+            }
+            return $subDirectory;
+        }
+
+        return $directory;
+    }
+
+    private function getScalesDirectory(core_kernel_classes_Resource $item): Directory
+    {
+        return $this
+            ->itemsService
+            ->getItemDirectory($item)
+            ->getDirectory(self::SCALES_DIRECTORY);
+    }
+
+    private function sanitizeRelativePath(string $path): string
+    {
+        $trimmed = trim($path);
+        $trimmed = str_replace('\\', '/', $trimmed);
+        $trimmed = preg_replace('#(\.\./|\.\\\)#', '', $trimmed);
+
+        return ltrim($trimmed, '/');
```

Comment:
lets add some security stuff because those paths come from user input. 
For sure we need to check if path is in scales directory and if its json file
` if (!str_starts_with($trimmed, self::SCALES_DIRECTORY . '/')) {
         throw new \InvalidArgumentException('Invalid path');
     }
`

`      if (!str_ends_with($trimmed, '.json')) {
          throw new \InvalidArgumentException('Must be a .json');
      }
`